### PR TITLE
chore: add video

### PR DESCRIPTION
**Pull request by koen1711**

### CODE REVIEW

**Comment by MartenPostma** [comment]
Reviewed the branch and pushed two commits with fixes; summary below so nothing lands in your branch unannounced.

**`ff6c54b` — ci: restore the workflow versions the merge rolled back.** The master merge resolved `.github/workflows/` toward the older side: `checkout@v5→v4`, `setup-python@v6→v5` in four workflows, and `--durations=10` dropped from the test run. All four files are back to master's version verbatim.

**`2b1e616` — fix(video): defer autoplay, respect click ownership, guard edge cases.** Four findings, each now covered by a test:

1. **`autoplay=True` made `when_video_starts` unreachable.** The event fired during `__init__`, before a callback can possibly be registered, and `started_once` then swallowed it forever — verified empirically before fixing. Autoplay is now deferred to the first `_tick` (one frame), so the obvious pattern — construct, then decorate — works. `play()`/`stop()`/`close()` cancel a pending autoplay.
2. **The controls bypassed the click-ownership system.** They keyed on the global `click_happened`, so a widget drawn over the video handled a click *and* the video toggled play underneath it. The click paths now go through `mouse_state.click_hits(self)`.
3. **Keyboard shortcuts ignored `controls=False`.** They ran on hover alone, so a game using Space would pause any video the mouse happened to be over. Now gated on `controls` as well.
4. **A stream without duration metadata (probe → 0.0) "ended" on its first playing frame** because the clock clamped to length. Zero length now means "unknown": the clock runs free and the end check is skipped.

Plus a drive-by: `_next_decoded` caught bare `Exception` where it meant `queue.Empty`.

Verified on the branch: 1250 passed, 1 skipped, 1 xfailed (was 1245); black clean; pylint 10.00/10.

**Noted but deliberately left to you**, since each is a design call:

- Scrubbing while playing rebuilds a `pygame.mixer.Sound` from the remaining PCM on every dragged frame — potentially tens of MB per drag frame. Restarting audio only on drag-release would cut that.
- Two overlapping videos still both respond to one click; `click_hits` only defers to widgets. Making Video participate in ownership resolution would change how it layers over plain sprites, so I didn't decide that here.
- `_mixer_formats` maps signed 8-bit mixers to s16 and >2-channel mixers to stereo, which would garble audio on those (rare) configurations.
- The suite now emits 66 `FutureWarning`s from `@experimental` on Video; a `filterwarnings` entry in `pyproject.toml` would keep the test output quiet if you want it.

Same note as on #183: `new_video` is public API, so play-docs needs a matching page in this change round.

The architecture reads well — the decoder generation counter, the state living off the sprite, and the `do_events` gating around the per-frame hooks are all exactly right.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: play/api/__init__.py [hunk 1]
```diff
@@ -33,6 +34,11 @@
     while_mouse_pressed,
     when_mouse_clicked,
     when_click_released,
+    when_video_ends,
+    when_video_starts,
+    when_video_plays,
+    when_video_pauses,
+    when_video_frame_changes,
```
Comment (MartenPostma) [inline play/api/__init__.py]: interesting, but where would one use this?

File: play/objects/video_controls.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+
+def _font(video):
+    player = video._player
+    if player.font is None:
+        try:
+            if not pygame.font.get_init():
+                pygame.font.init()
+            player.font = pygame.font.SysFont(None, max(11, video._bar_height() // 3))
+        except Exception:  # pylint: disable=broad-except  # pragma: no cover
```
Comment (MartenPostma) [inline play/objects/video_controls.py]: except Exception, is there a more specific error type?

File: play/objects/video.py [hunk 3]
```diff
@@ -0,0 +1,940 @@
...
+
+        self.hover = False
+        self.controls_alpha = 0.0
+        self.last_activity = 0.0
+        self.scrubbing = False
+        self.dragging_volume = False
+        self.consumed_click = False
+
+
+@experimental
```
Comment (MartenPostma) [inline play/objects/video.py]: do we want the @experimental decorator for sound and video?

**Review by MartenPostma — CHANGES_REQUESTED** [review]

File: play/api/__init__.py (hunk 1 above)
Comment (koen1711) [reply play/api/__init__.py]: Where one wants feedback on what an user does with an video, maybe they want to show something when the video pauses?

File: play/objects/video.py (hunk 3 above)
Comment (koen1711) [reply play/objects/video.py]: debateable

File: play/objects/video_controls.py (hunk 2 above)
Comment (koen1711) [reply play/objects/video_controls.py]: this is quite common among font code and other loading of external files throughout the project, we could improve this in another PR however?

**Review by MartenPostma — APPROVED** [review]
